### PR TITLE
197 225 get chat messages and delete conversations

### PR DESCRIPTION
**Pull request by wiam8**
This PR tackles issues #225 and #197. 

Previously, the listeners to the conversations were added on the ChatActivity using the list listener, which wasn't technically correct. A helper can start messaging a second user, and those texts would be simply lost if the user hadn't opened the chat activity yet as the list listener only allowed getting the last added value. The same thing applies to the users who replied to the helpRequest.
This was corrected by using the Child Event Listener of firebase which is triggered for every child added. The child event listener is also triggered when a child is removed from the location to which this listener was added, which allowed me to also implement the deletion of conversations and conversation ids. 

Initially, I wanted to change the value Listener to a child Event Listener in our code. However, after investigating this, I realized that this would lead to MAJOR changes in our current codebase as I would have to basically rewrite all the listener's methods of the database files and function calls, so I chose instead to include this as an additional method for the time being. I still think that this is better as in most listeners we are using so far, we don't need to detect the removal of a child, and so if we decide to use child Event Listeners we would have to pass null arguments multiples times. 

The changes made include the following: 
- Every user who responds to the help request is added to the RecentMessagesActivity under Helper + index, the helpee can thus start communicating with him. If the helper decides later on that he can't provide help, he can cancel. The conversation is thus deleted from the database, the helper is redirected to the main page and the helpee is redirected to the RecentMessagesActivity. A message explaining that the helper has canceled is displayed. The conversation is also removed from the recentMessagesActivity. 
- If however, the helpee cancels his emergency, every helper who previously agreed to help is redirected to the main page, all the conversations are deleted from the database and the conversation ids of this user are also deleted. The helpee is also redirected to the main page after he cancels his emergency.

### CODE REVIEW

File: app/src/main/java/com/github/h3lp3rs/h3lp/messaging/ChatActivity.kt [hunk 1]
```diff
@@ -38,28 +44,85 @@ class ChatActivity : AppCompatActivity() {
...
+     * @param message message to display
+     */
+    private fun displayMessage(message : String) {
+        Toast.makeText(
+            applicationContext, message, Toast.LENGTH_LONG
+        ).show()
+    }
+
+    /**
+     * Send back to the MainPageActivity
```
Comment (Lindt19) [inline app/src/main/java/com/github/h3lp3rs/h3lp/messaging/ChatActivity.kt:121]:
```suggestion
     * Sends back to the MainPageActivity
```

File: app/src/main/java/com/github/h3lp3rs/h3lp/messaging/RecentMessagesActivity.kt [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    private var idToItem = Collections.synchronizedMap<String, HelperConversation>(mutableMapOf())
+
+    override fun onCreate(savedInstanceState : Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_latest_messages)
+
+        val bundle = this.intent.extras
+        helpeeId = bundle?.getString(EXTRA_HELPEE_ID) ?: helpeeId
+
+        databaseOf(Databases.MESSAGES).setString("HEWIEWEWE", helpeeId!!)
```
Comment (Lindt19) [inline app/src/main/java/com/github/h3lp3rs/h3lp/messaging/RecentMessagesActivity.kt:42]: You can directly use the messagesDatabase attribute (Totally understood the message string 😜)

File: app/src/main/java/com/github/h3lp3rs/h3lp/messaging/RecentMessagesActivity.kt [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+     * @param message message to display
+     */
+    private fun displayMessage(message : String) {
+        Toast.makeText(
+            applicationContext, message, Toast.LENGTH_LONG
+        ).show()
+    }
+
+    private fun listenForRemovedHelpers() {
+        // Event listener that handles deleting the conversation from the view upon deletion
```
Comment (Lindt19) [inline app/src/main/java/com/github/h3lp3rs/h3lp/messaging/RecentMessagesActivity.kt:93]: Just for consistency, maybe you can use the same documentation style as the onChildAdded method above with the @param field

**Review by Lindt19 — APPROVED** [review]
Very clean implementation of the delete operation! Keep up the good work 🚀

File: app/src/main/java/com/github/h3lp3rs/h3lp/messaging/RecentMessagesActivity.kt (hunk 2 above)
Comment (wiam8) [reply app/src/main/java/com/github/h3lp3rs/h3lp/messaging/RecentMessagesActivity.kt:42]: Haha agree, but that wasn't done by me, it should be corrected in #201

File: app/src/main/java/com/github/h3lp3rs/h3lp/HelpPageActivity.kt [hunk 4]
```diff
@@ -216,4 +223,23 @@ class HelpPageActivity : AppCompatActivity(), CoroutineScope by MainScope() {
     private fun goToMainPage() {
         goToActivity(MainPageActivity::class.java)
     }
+
+    private fun onEmergencyCancelled() {
```
[the remark this answers is not in the document]
Comment (wiam8) [reply app/src/main/java/com/github/h3lp3rs/h3lp/HelpPageActivity.kt:227]: Good point!

File: app/src/main/java/com/github/h3lp3rs/h3lp/HelpPageActivity.kt [hunk 5]
```diff
@@ -216,4 +223,23 @@ class HelpPageActivity : AppCompatActivity(), CoroutineScope by MainScope() {
...
+                goToActivity(MainPageActivity::class.java)
+            }
+        }
+        // The event is added to the entire conversation IDS database and so no child key is needed
+        conversationIdsDb.addEventListener(
+            null,
+            String::class.java, null,
+        ) { id -> run { onChildRemoved(id) } }
+    }
+
```
[the remark this answers is not in the document]
Comment (wiam8) [reply app/src/main/java/com/github/h3lp3rs/h3lp/HelpPageActivity.kt:243]: I'm not sure about this. Don't we want the user to still have access to the conversation even if he accidentally quits the app or his phone dies? The way I see it is that the conversation should still be saved unless the user explicitly chooses to delete it. In the case where user A quits the app, user B can still delete the conversation from the database if he stops getting messages from user A. 
However, the help page activity (and thus the chat activity for the helper) is only launched from the notification and so the conversations wouldn't be accessible after quitting the app. This is probably something we should fix in another sprint

File: app/src/main/java/com/github/h3lp3rs/h3lp/database/Database.kt [hunk 6]
```diff
@@ -163,5 +163,20 @@ interface Database {
...
-    fun incrementAndGet(key: String, increment: Int, onComplete: (Int?) -> Unit)
+    fun incrementAndGet(key : String, increment : Int, onComplete : (Int?) -> Unit)
+
+    /**
+     * Adds an event listener to a key in the database
+     * @param onChildAdded The action to take on the new data snapshot when data is added. Could be
+     * null
+     * @param onChildRemoved The action to take on the new data snapshot when data is removed
+     * @param key The relative path from this reference to the new one or null if the action is to
+     * be taken on the entire original path
```
[the remark this answers is not in the document]
Comment (wiam8) [reply app/src/main/java/com/github/h3lp3rs/h3lp/database/Database.kt:174]: oops 👉🏻👈🏻

File: app/src/main/java/com/github/h3lp3rs/h3lp/HelpPageActivity.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (Ahayderi) [reply app/src/main/java/com/github/h3lp3rs/h3lp/HelpPageActivity.kt:239]: Yes I was considering the fact that after quitting the app, the helper wouldn't have access to the help page (or conversely for the helpee) but indeed, if we plan to improve this at a later point in time, keeping the code as is is fine.

**Pull request merged**